packages/flutter/lib/src/rendering/object.dart

```diff
@@ -953,18 +935,38 @@ class PipelineOwner {
 
   /// The object that is managing semantics for this pipeline owner, if any.
   ///
-  /// An owner is created by [addSemanticsListener] the first time a listener is
-  /// added.
-  ///
-  /// The owner is valid for as long as there are listeners. Once the last
-  /// listener is removed (by calling [SemanticsOwner.removeListener] on the
-  /// [semanticsOwner]), the [semanticsOwner] field will revert to null, and the
-  /// previous owner will be disposed.
+  /// An owner is created by [ensureSemantics]. The owner is valid for as long
+  /// there are [SemanticsHandle] returned by [ensureSemantics] that have not
+  /// yet be disposed. Once the last handle has been disposed, the
+  /// [semanticsOwner] field will revert to null, and the previous owner will be
+  /// disposed.
   ///
   /// When [semanticsOwner] is null, the [PipelineOwner] skips all steps
   /// relating to semantics.
   SemanticsOwner get semanticsOwner => _semanticsOwner;
   SemanticsOwner _semanticsOwner;
+
+  int _outstandingSemanticsHandle = 0;
+
+  SemanticsHandle ensureSemantics({ VoidCallback listener }) {
+    if (_outstandingSemanticsHandle++ == 0) {
+      _semanticsOwner = new SemanticsOwner();
+      final AbstractNode rootNode = _rootNode;
+      if (rootNode is RenderObject)
+        rootNode.scheduleInitialSemantics();
+    }
+    return new SemanticsHandle._(this, listener);
+  }
+
+  void _didDisposeSemanticsHandle() {
+    if (--_outstandingSemanticsHandle == 0) {
+      _semanticsOwner?.dispose();
+      final AbstractNode rootNode = _rootNode;
+      if (rootNode is RenderObject)
+        rootNode.clearSemantics();
```

Comment:
i think we need to go back to the crazy callback if we're going to continue to support non-render-object abstract node roots (which we probably should because it keeps us honest).